File: cipher-codes/AlbertiCipher.cpp
#include <iostream>
#include <unordered_map>
#include <string>

using namespace std;

unordered_map<char, char> createAlbertiWheelMapping(const string& outsideWheel, const string& insideWheel) {
    unordered_map<char, char> mapping;
    for (size_t i = 0; i < outsideWheel.length(); i++) {
        mapping[outsideWheel[i]] = insideWheel[i];
    }
    return mapping;
}

string encryptAlberti(const string& plaintext, const string& outsideWheel, const string& insideWheel) {
    unordered_map<char, char> mapping = createAlbertiWheelMapping(outsideWheel, insideWheel);
    string ciphertext;
    for (char c : plaintext) {
        char upperC = toupper(c);
        if (mapping.find(upperC) != mapping.end()) {
            ciphertext += mapping[upperC];
        } else {
            ciphertext += c;
        }
    }
    return ciphertext;
}

string decryptAlberti(const string& ciphertext, const string& outsideWheel, const string& insideWheel) {
    unordered_map<char, char> reverseMapping = createAlbertiWheelMapping(insideWheel, outsideWheel);
    string plaintext;
    for (char c : ciphertext) {
        if (reverseMapping.find(c) != reverseMapping.end()) {
            plaintext += reverseMapping[c];
        } else {
            plaintext += c;
        }
    }
    return plaintext;
}

int main() {
    string outsideWheel, insideWheel, mode, text;

    cout << "Enter the characters of the outside wheel: ";
    getline(cin, outsideWheel);
    for (char& c : outsideWheel) c = toupper(c);

    cout << "Enter the characters of the inside wheel: ";
    getline(cin, insideWheel);
    for (char& c : insideWheel) c = toupper(c);

    if (outsideWheel.length() != insideWheel.length()) {
        cout << "Error: The outside and inside wheels must have the same length." << endl;
        return 1;
    }

    cout << "Do you want to encrypt or decrypt? (e/d): ";
    getline(cin, mode);

    if (mode == "e") {
        cout << "Enter the plaintext message: ";
        getline(cin, text);
        string encryptedText = encryptAlberti(text, outsideWheel, insideWheel);
        cout << "Encrypted message: " << encryptedText << endl;
    } else if (mode == "d") {
        cout << "Enter the encrypted message: ";
        getline(cin, text);
        string decryptedText = decryptAlberti(text, outsideWheel, insideWheel);
        cout << "Decrypted message: " << decryptedText << endl;
    } else {
        cout << "Invalid option. Please choose 'e' to encrypt or 'd' to decrypt." << endl;
    }

    return 0;
}

File: cipher-codes/AlbertiCipher.py
def create_alberti_wheel_mapping(outside_wheel, inside_wheel):
    return {outside_wheel[i]: inside_wheel[i] for i in range(len(outside_wheel))}

def encrypt_alberti(plaintext, outside_wheel, inside_wheel):
    mapping = create_alberti_wheel_mapping(outside_wheel, inside_wheel)
    ciphertext = ''.join([mapping[char.upper()] if char.upper() in mapping else char for char in plaintext])
    return ciphertext

def decrypt_alberti(ciphertext, outside_wheel, inside_wheel):
    reverse_mapping = create_alberti_wheel_mapping(inside_wheel, outside_wheel)
    plaintext = ''.join([reverse_mapping[char] if char in reverse_mapping else char for char in ciphertext])
    return plaintext

def main():
    outside_wheel = input("Enter the characters of the outside wheel: ")
    inside_wheel = input("Enter the characters of the inside wheel: ")

    if len(outside_wheel) != len(inside_wheel):
        print("Error: The outside and inside wheels must have the same length.")
        return

    mode = input("Do you want to encrypt or decrypt? (e/d): ").strip().lower()

    if mode == 'e':
        plaintext = input("Enter the plaintext message: ")
        encrypted_text = encrypt_alberti(plaintext, outside_wheel, inside_wheel)
        print("Encrypted message:", encrypted_text)
    elif mode == 'd':
        ciphertext = input("Enter the encrypted message: ")
        decrypted_text = decrypt_alberti(ciphertext, outside_wheel, inside_wheel)
        print("Decrypted message:", decrypted_text)
    else:
        print("Invalid option. Please choose 'e' to encrypt or 'd' to decrypt.")

if __name__ == "__main__":
    main()
